### ISSUE TITLE
Trimming known adapter pairings

### ISSUE THREAD

**Issue by marcelm** [marcelm/cutadapt#44]
_From [emur...@gmail.com](https://code.google.com/u/116561790744379098758/) on December 16, 2011 04:18:57_

This is not a bug, just a request.  To illustrate the issue, lets pretend:

1. I have a sequence to be trimmed that looks like this:
   5' TTGsomenucleotides-adapter2 3'

2. My adapter sequences are:

   adapter1
   adapter2, which ends in TTG on the 3' end

3. All possible sequenced fragments generated for sequencing will be flanked with Adapter1+Adapter2, or RCAdapter1+RCAdapter2.  No authentic fragment will ever be flanked by Adapter2 on both sides, or Adapter1 on both sides.  This is the case with the Illumina TruSeq protocols, if I'm not mistaken; it may be common with many protocols.

In this scenario, 5' trimming of the file using all 4 potential adapters would mean that the example sequence would have the beginning "TTG" removed erroneously.  We can be sure it is not the end of Adapter2, even though it might match, because Adapter2 will never flank the same fragment on both sides.

A single -b (or -a or -g) argument cannot be relied upon, because other sequences in the same file might have "real" adapter sequences on both ends.  However, two iterations of -b will again erroneously remove this portion of "real" sequence after having removed the better match at the 3' end in the first iteration.

Proposed solution(s):

It would be useful if a user could enter adapters as linked pairs.  The algorithm which currently decides which of all possible single adapters is most valid could weigh which of the "pairs" was most likely occurring based on combined match length of both adapters of a pair.  Then, strong/certain matches of an adapter at one end could prevent the opposite end from being erroneously trimmed using an adapter match that should not occur in the reaction.  This implementation could also provide two useful mechanisms for screening adapter concatenations from real organism sequences with minimal loss of true sequence without (I think) over-encumbering the program.

First, a person could run the program using pairs that should not occur flanking a "real" fragment (for example Adapter1+Adapter1), set the -O very high to make sure only near-complete adapters were recognized, and instruct the program to discard reads matching these criteria rather than trimming the reads.  

A second mechanism is enabled during normal trimming by the match comparison for deciding which pair is correct to trim a given sequence.  For example, I could enter 2 pairs of adapter sequences, each 30 bp in length: 1&2 occur together, and 3&4 occur together.  Normally, (I think) your program would find the best match or, in the case of equally good matches, choose the first pair.  A second length parameter (-OO or something) could be set so that if a "tie" occurred with both pairs matching along at least that length, that read would be discarded rather than trimmed using the first pair.  In my example, if I set -OO 28, any "ties" in best pair decision resulting from a sequence matching 28 bp of adapter1 AND 28 bp of adapter3 would cause a sequence to be discarded (or output to a separate file).

In summary, I think the ability to input/compare adapter "pairs" could help reduce loss of authentic data at high stringency while enabling quick identification and removal of "junk" sequences through mechanisms you have already set in place.

Lastly, thank you for your work creating, updating, and improving this tool.  It is really fantastic and is one of very few bioinformatic tools that permits immediate, flexible, and accurate use with little background computer knowledge demanded from the user and only a single dependency.  I cannot express how much time and frustration your tool has saved me, and how much I appreciate and admire your work.

Sincerely,
Elspeth Murday
Clemson University
emurday@g.clemson.edu

_Original issue: http://code.google.com/p/cutadapt/issues/detail?id=34_

**Comment by marcelm** [comment]
_From [marcel.m...@tu-dortmund.de](https://code.google.com/u/115749445199231802729/) on December 16, 2011 05:36:17_

Dear Elspeth Murday,

before I forget: thanks for the kind words in your last paragraph! It's great to hear the work is being appreciated.

I think your request makes a lot of sense. This is a use case I had not thought about, but it sounds as if it would be nice to support this in cutadapt. The problem is that I'm currently busy with a lot of other non-cutadapt things, so this may take a while. Please be patient or try to get someone else to implement this - the Python source code shouldn't be too hard too understand.

Marcel

**Labels:** -Type-Defect Type-Enhancement

**Comment by marcelm** [comment]
_From [kbh...@gmail.com](https://code.google.com/u/105267345598351240433/) on July 23, 2012 08:29:34_

We would also find it very useful to cut on both sides at the same time.  And even better, if per the description above, if it avoids any erroneous cutting because it has both adaptor sequences to begin with.  Likewise, your work so far is very very solid and appreciated, thank you!

**Comment by marcelm** [comment]
This has been implemented as "linked adapters" and will be part of the next release of cutadapt.
